File: Crystal/Math/Circle3d.cpp
#include "Circle3d.h"

#include "Tolerance.h"
#include <cmath>

using namespace Crystal::Math;

Circle3d::Circle3d() :
	center(Vector3dd(0,0,0)),
	uvec(Vector3dd(1,0,0)),
	vvec(Vector3dd(0,1,0))
{}

Circle3d::Circle3d(const Vector3dd& center, const Vector3dd& uvec, const Vector3dd& vvec) :
	center(center),
	uvec(uvec),
	vvec(vvec)
{}

Vector3dd Circle3d::getPosition(const double u) const
{
	const auto uu = u * 2.0 * PI;
	return center + ::cos(uu) * uvec + ::sin(uu) * vvec;
}

Vector3dd Circle3d::getNormal() const
{
	return glm::normalize( glm::cross(uvec, vvec) );
}

File: SandBox/CS/PGCore/Math/Primitive3d/Circle3d.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PG.Core.Math
{
    public class Circle3d : ICurve3d, ISurface3d
    {
        public double Radius { get; }

        public Vector3d Center { get; }
            = new Vector3d(0, 0, 0);

        public Circle3d() :
            this(1.0, new Vector3d(0,0,0))
        { }

        public Circle3d(double radius, Vector3d center)
        {
            this.Radius = radius;
            this.Center = center;
        }

        public Vector3d GetPosition(double u)
        {
            return GetPosition(u, 1.0);
        }

        public Vector3d GetPosition(double u, double v)
        {
            var angle = u * 2.0 * System.Math.PI;
            var x = v * Radius * System.Math.Cos(angle);
            var y = v * Radius * System.Math.Sin(angle);
            return Center + new Vector3d(x, y, 0.0);
        }

        public bool IsSame(Circle3d rhs, double tolerance)
        {
            return this.Center.IsSame(rhs.Center, tolerance) &&
                System.Math.Abs(this.Radius - rhs.Radius) < tolerance;
        }

    }
}
